interfaces/builtin/gpio.go

```diff
@@ -0,0 +1,150 @@
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/snapcore/snapd/interfaces"
+)
+
+var gpioSysfsGpioBase = "/sys/class/gpio/gpio"
+var gpioSysfsExport = "/sys/class/gpio/export"
+
+// GpioInterface type
+type GpioInterface struct{}
+
+// String returns the same value as Name().
+func (iface *GpioInterface) String() string {
+	return iface.Name()
+}
+
+// Name of the GpioInterface
+func (iface *GpioInterface) Name() string {
+	return "gpio"
+}
+
+// SanitizeSlot checks the slot definition is valid
+func (iface *GpioInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	// Paranoid check this right interface type
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	// We will only allow creation of this type of slot by a gadget or OS snap
+	if !(slot.Snap.Type == "gadget" || slot.Snap.Type == "os") {
+		return fmt.Errorf("gpio slots only allowed on gadget or core snaps")
+	}
+
+	// Must have a GPIO number
+	number, ok := slot.Attrs["number"]
+	if !ok {
+		return fmt.Errorf("gpio slot must have a number attribute")
+	}
+
+	// Valid values of number
+	if _, ok := number.(int); !ok {
+		return fmt.Errorf("gpio slot number attribute must be an int")
+	}
+
+	// Slot is good
+	return nil
+}
+
+// SanitizePlug checks the plug definition is valid
+func (iface *GpioInterface) SanitizePlug(plug *interfaces.Plug) error {
+	// Make sure right interface type
+	if iface.Name() != plug.Interface {
+		panic(fmt.Sprintf("plug is not of interface %q", iface))
+	}
+
+	// Plug is good
+	return nil
+}
+
+// PermanentPlugSnippet returns security snippets for plug at install
+func (iface *GpioInterface) PermanentPlugSnippet(plug *interfaces.Plug, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityDBus, interfaces.SecurityAppArmor, interfaces.SecuritySecComp, interfaces.SecurityUDev, interfaces.SecurityMount:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+// ConnectedPlugSnippet returns security snippets for plug at connection
+func (iface *GpioInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityDBus, interfaces.SecuritySecComp, interfaces.SecurityUDev, interfaces.SecurityMount:
+		return nil, nil
+	case interfaces.SecurityAppArmor:
+		path := fmt.Sprint(gpioSysfsGpioBase, slot.Attrs["number"])
+		// Entries in /sys/class/gpio for single GPIO's are just symlinks
+		// to their correct device part in the sysfs tree. As AppArmor
+		// does not handle symlinks we need to dereference the GPIO
+		// path and add the correct absolute path to the AppArmor snippet.
+		dereferencedPath, err := evalSymlinks(path)
+		if err != nil {
+			return nil, err
+		}
+		return []byte(fmt.Sprintf("%s/* rwk,\n", dereferencedPath)), nil
```

Comment:
I'm not sure why 'k' (lock) is needed here, but I see it in bool-file, so I guess that's ok. The security policy LGTM-- it is basically this one rule per pin for connected plugs, which is equivalent to what we have in bool-file. Interestingly, because the slot side is limited to the core snap (unconfined) or the gadget snap (currently unconfined), there are no needed slot permissions. Note that if we expect the gadget snap to be confined at some point, we'll need to have permanent slot policy conditional on if gadget. A TODO or NOTE comment in this PR describing this is welcome but not a blocker.

+1 for security policy